src/VisualStudio/Core/Def/Utilities/VsCodeWindowViewTracker.cs

```diff
@@ -0,0 +1,139 @@
+    /// tracking the lifetime of the views as appropriate.
+    /// </summary>
+    /// <remarks>
+    /// All members of this class are UI thread affinitized, including the constructor.
+    /// </remarks>
+    internal sealed class VsCodeWindowViewTracker : ForegroundThreadAffinitizedObject, IDisposable, IVsCodeWindowEvents
+    {
+        private readonly IVsCodeWindow _codeWindow;
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+
+        private readonly ComEventSink _codeWindowEventsSink;
+
+        /// <summary>
+        /// The map from <see cref="IVsTextView"/> and corresponding <see cref="ITextView"/> for views that we are currently watching for caret movements.
+        /// </summary>
+        private readonly Dictionary<IVsTextView, ITextView> _trackedTextViews = new();
+
+        public VsCodeWindowViewTracker(IVsCodeWindow codeWindow, IThreadingContext threadingContext, IVsEditorAdaptersFactoryService editorAdaptersFactoryService)
+            : base(threadingContext, assertIsForeground: true)
+        {
+            _codeWindow = codeWindow;
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+
+            _codeWindowEventsSink = ComEventSink.Advise<IVsCodeWindowEvents>(codeWindow, this);
+
+            if (ErrorHandler.Succeeded(_codeWindow.GetPrimaryView(out var pTextView)) && pTextView != null)
+                StartTrackingView(pTextView);
+
+            // If there is no secondary view, GetSecondaryView will return null
+            if (ErrorHandler.Succeeded(_codeWindow.GetSecondaryView(out pTextView)) && pTextView != null)
+                StartTrackingView(pTextView);
+        }
+
+        private void StartTrackingView(IVsTextView pTextView)
+        {
+            AssertIsForeground();
+
+            if (!_trackedTextViews.ContainsKey(pTextView))
+            {
+                var wpfTextView = _editorAdaptersFactoryService.GetWpfTextView(pTextView);
+
+                if (wpfTextView != null)
+                {
+                    _trackedTextViews.Add(pTextView, wpfTextView);
+                    wpfTextView.Caret.PositionChanged += OnCaretPositionChanged;
+                    wpfTextView.GotAggregateFocus += OnViewGotAggregateFocus;
+                }
+            }
+        }
+
+        private void StopTrackingView(IVsTextView pView)
+        {
+            AssertIsForeground();
+
+            if (_trackedTextViews.TryGetValue(pView, out var view))
+            {
+                view.Caret.PositionChanged -= OnCaretPositionChanged;
+                view.GotAggregateFocus -= OnViewGotAggregateFocus;
+
+                _trackedTextViews.Remove(pView);
+            }
+        }
+
+        public ITextView GetActiveView()
+        {
+            AssertIsForeground();
+
+            ErrorHandler.ThrowOnFailure(_codeWindow.GetLastActiveView(out var pView));
```

Comment:
slightly scary throwing here, instead of just making this a Try.

Reply:
The previous code would have thrown exceptions in this case. There's no reasonable thing to ask the caller to do with a null other than silently march on, but I'm not aware of issues here so it's probably fine to keep as is. I agree if there is a case we need to care about, a Try pattern would be wise.

Reply:
wfm